spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ScopePassingSpanSubscriber.java

```diff
@@ -38,69 +40,71 @@
 
 	private static final Log log = LogFactory.getLog(ScopePassingSpanSubscriber.class);
 
-	private final Span span;
-
 	private final Subscriber<? super T> subscriber;
 
 	private final Context context;
 
-	private final Tracer tracer;
+	private final Tracing tracing;
+	private final CurrentTraceContext currentTraceContext;
+	private final TraceContext root;
 
 	private Subscription s;
 
 	ScopePassingSpanSubscriber(Subscriber<? super T> subscriber, Context ctx,
 			Tracing tracing) {
 		this.subscriber = subscriber;
-		this.tracer = tracing.tracer();
-		Span root = ctx != null ? ctx.getOrDefault(Span.class, this.tracer.currentSpan())
-				: null;
-		this.span = root;
-		this.context = ctx != null && root != null ? ctx.put(Span.class, root)
+		this.tracing = tracing;
+		this.currentTraceContext = this.tracing.currentTraceContext();
+		TraceContext root = ctx != null ? ctx.getOrDefault(TraceContext.class, this.currentTraceContext.get()) : null;
```

Comment:
I wonder if this is a little inefficient. `currentTraceContext.get()` will be called even if there is a TraceContext there.. is there a way to pass a method reference instead? otherwise, maybe get if it is null